CUSTOM.md

```diff
@@ -370,3 +370,11 @@ This section is meant for users who are familiar with both the tts mod and progr
         - 4: Slow Phase
         - 5: Time Passes
       - `turn`: **number** - is the current turn number
+- Trigger to run when a player gains/pays energy (or undoes these operations)
+  - Create object and tag with "Gain Pay"
+  - `onGainPay(params)`
```

Comment:
should this api have a return value? i imagine there could be a world where someone wants to modify the results. like maybe for Back Against the Wall or i guess a different way to implement bargain scripting

Reply:
That's a great idea for *gaining* energy. It'd certainly work for Back Against the Wall (I'll probably make that part of this PR), and I think it could work for Bargains (I'll need to think a bit more about that).

If it could also work for *paying* energy, that'd be fantastic, 'cause it'd solve one of the big problems with implemented Dances Up Earthquakes: cards go into play when you *gain* energy, but then should be ignored when you *pay* energy. But where it's currently called it can't really work for paying energy, because if you don't have energy energy, clicking the pay button will fail before it calls the hook. And, more problematically, this couldn't adjust the *displayed* total cost of cards. Anything to deal with modifying the cost you pay would thus have to hook in rather deeper.

Allowing a return values when gaining energy but not when paying energy would seem weird when it's one `onGainPay()` function. That would seem to merit splitting it into one `onGain()` and one `onPay()` function.

Thoughts on splitting it, and/or on hooking in something deeper to change the calculated cost to pay?

Reply:
So to be clear I'm not saying that we should redo how bargains are programmed, but I can imagine a world where we could have done it this way to make sure the API is robust enough to handle that potential feature

Reply:
Hmm I wonder if the hookup for paying should be earlier in the function? Another thought experiment would be for blitz scenario which modifies what you pay for cards and what not

Reply:
*Just* putting the hook for paying earlier in the function doesn't quite work, for two reasons:
- You really need two hooks. One gets called early in the function and is given the opportunity to modify the cost to pay. It then returns control to the global script, which checks if you *can* pay that cost, and only *if* you can, calls the second hook, which does the actual on-pay stuff.
- Even with the two-hook model I describe there, the constantly-displayed "Energy Cost" label is wrong, so really the first hook would have to go into the update loop that manages that label.

Reply:
Thinking about it further, I think a full implementation of pay/gain with cost modification looks like this:

Reply:
On further consideration, I think a full implementation allowing modifying the cost of gain/pay would look like this:
- Trigger to modify the cost of playing cards
  - Create object and tag with "Modify Cost"
  - `modifyCost(params)`
    -`params`: **table** - contains data about the player and the cards played
      - `color`: **string** - is the color of the player whose card cost is being calculated
      - `costs`: **table** - maps card GUIDs (as keys) to costs (as values)
    - return **table** - the `costs` table, modified appropriately
      - e.g. Blitz would reduce the cost of all fast cards by 1, cards could be ignored in cost calculations by deleting them from the table
- Trigger to run when a player gains energy (or undoes this operation)
  - Create object and tag with "Gain Energy"
  - `onGain(params)`
    - `params`: **table** - contains data about the gain operation
      - `color`: **string** - is the color of the player who gained energy
      - `isUndo`: **bool** - true if the operation is being undone (right click), false otherwise
      - `amount`: **number** - amount of energy gained (before bargain deductions)
    - `return`: **number** - the amount of energy to gain
- Trigger to run when a player pays energy (or undoes this operation)
  - Create object and tag with "Pay Energy"
  - `onPay(params)`
    - `params`: **table** - contains data about the pay operation
      - `color`: **string** - is the color of the player who paid energy
      - `isUndo`: **bool** - true if the operation is being undone (right click), false otherwise
      - `amount`: **number** - amount of energy paid
`modifyCost()` would hook into the `countItems()` loop set up by `setupPlayerArea()`.

A Dances Up Earthquakes implementation would then ignore, for the purposes of cost calculations, any cards that had come into pay from impending this turn. Then, even if Blitz were implemented through this same interface, it wouldn't matter in which order the Blitz `modifyCost()` and the Dances `modifyCost()` were called; it'd work either way.

I'm happy to write this up, but I realise this is more complication than you might be willing to add in. Thoughts?

Reply:
so i'm not sure that DUE should have special handling for having cards that are no longer impending ignored in energy cost since if a spirit is allowed to play extra cards or pull cards back from in play they currently change what the energy cost label says

Reply:
The reason Dances would need it is that the natural order is to hit the Gain button before the Pay button (and you often need to do it in that order to have enough energy to pay), and gaining energy is when cards come into play from impending. An alternative implementation *could* be to have cards not go into play from impending until you hit Pay, but the problem with that idea is that you don't have all the elements from your impending cards when you're choosing which cards to play this turn.

But honestly, that's all slightly beside the point. Even if we just work with the thought experiment of Blitz that you suggested, that requires changing how the energy cost label is calculated, necessitating the existence of something like the `modifyCost()` function.

Reply:
in general i think it's fine to proceed down this road with the extra hookups and what not